Update documentation of ml inference processors to support for local models (#7368)
opensearch-project/documentation-website: _ingest-pipelines/processors/ml-inference.md

The `ml_inference` processor is used to generate inferences from machine learning (ML) models connected to the [OpenSearch ML Commons plugin]({{site.url}}{{site.baseurl}}/ml-commons-plugin/). The inferences are added as new fields to the ingested documents.

**PREREQUISITE**<br>
Before using the `ml_inference` processor, you must have either a local ML model hosted on your OpenSearch cluster or a remote model connected to your OpenSearch cluster through the ML Commons plugin.

For more information on remote models, see [Connecting to externally hosted models]({{site.url}}{{site.baseurl}}/ml-commons-plugin/remote-models/index/).

> For more information on remote models, see [Using ML models within OpenSearch]({{site.url}}{{site.baseurl}}/ml-commons-plugin/using-ml-models/).

Reviewer: For more information on local models,

Author: good catch, thanks

> For more information on remote models, see [Using ML models within OpenSearch]({{site.url}}{{site.baseurl}}/ml-commons-plugin/using-ml-models/).

Reviewer 2: Local model ?

    ],
    "model_config":{
      "<model_config_field>": "<config_value>"
    },

>     "override": "<override>"

Reviewer 2: minor: move this config parameter to after line 31 ?

The following table lists the required and optional parameters for the `ml-inference` processor.

| Parameter | Data type | Required/Optional | Description |
|:--- | :--- | :--- | :--- |

> | `model_id` | String | Required | The ID of the ML model to be used by the processor. |

Reviewer 3: ```suggestion
| `model_id` | String | Required | The ID of the ML model used by the processor. |
```

| `model_input`  | String    | Optional for externally hosted models<br/><br/>Required for local models | A template that defines the input field format expected by the model. Each local model type might use a different set of inputs. For externally hosted models, default is `"{ \"parameters\": ${ml_inference.parameters} }`.|
| `input_map` | Array | Optional for externally hosted models<br/><br/>Required for local models | An array specifying how to map ingested document fields to the model input fields. Each element of the array is a map in the `"<model_input_field>": "<document_field>"` format and corresponds to one model invocation for a document field. If no input mapping is specified for an externally hosted model, then all fields from the document are passed to the model directly as input. The `input_map` size indicates the number of times the model is invoked (the number of Predict API requests). |
| `<model_input_field>` | String    | Optional for externally hosted models<br/><br/>Required for local models  | The model input field name. |

> | `<document_field>`   | String    | Optional for externally hosted models<br/><br/>Required for local models | The name or JSON path of the ingested document field to be used as the model input. |

Reviewer 3: ```suggestion
| `<document_field>`   | String    | Optional for externally hosted models<br/><br/>Required for local models | The name or JSON path of the ingested document field used as the model input. |
```

| `output_map` | Array | Optional for externally hosted models<br/><br/>Required for local models | An array specifying how to map the model output fields to new fields in the ingested document. Each element of the array is a map in the `"<new_document_field>": "<model_output_field>"` format.|

> | `<new_document_field>`   | String    | Optional for externally hosted models<br/><br/>Required for local models | The name of the new field in the ingested document in which the model's output (specified by the `model_output`) is stored. If no output mapping is specified for externally hosted models, then all fields from the model output are added to the new document field. |

Reviewer 3: ```suggestion
| `<new_document_field>`   | String    | Optional for externally hosted models<br/><br/>Required for local models | The name of the new field in the ingested document in which the model's output (specified by `model_output`) is stored. If no output mapping is specified for externally hosted models, then all fields from the model output are added to the new document field. |
```

| `<model_output_field>` | String    | Optional for externally hosted models<br/><br/>Required for local models | The name or JSON path of the field in the model output to be stored in the `new_document_field`. |
| `full_response_path` | Boolean   | Optional   | Set this parameter to `true` if the `model_output_field` contains a full JSON path to the field instead of the field name. The model output will then be fully parsed to get the value of the field. Default is `true` for local models and `false` for externally hosted models. |
| `ignore_missing` | Boolean   | Optional  | If `true` and any of the input fields defined in the `input_map` or `output_map` are missing, then the missing fields are ignored. Otherwise, a missing field causes a failure. Default is `false`. |

> | `ignore_failure` | Boolean   | Optional  | Specifies whether the processor continues execution even if it encounters errors. If `true`, any failure is ignored and ingestion continues. If `false`, any failure causes ingestion to be canceled. Default is `false`. |

Reviewer 3: ```suggestion
| `ignore_failure` | Boolean   | Optional  | Specifies whether the processor continues execution even if it encounters an error. If `true`, then any failure is ignored and ingestion continues. If `false`, then any failure causes ingestion to be canceled. Default is `false`. |
```

> | `override` | Boolean   | Optional   | Relevant if an ingested document already contains a field with the name specified in `<new_document_field>`. If `override` is `false`, the input field is skipped. If `true`, the existing field value is overridden by the new model output. Default is `false`. |

Reviewer 3: ```suggestion
| `override` | Boolean   | Optional   | Relevant if an ingested document already contains a field with the name specified in `<new_document_field>`. If `override` is `false`, then the input field is skipped. If `true`, then the existing field value is overridden by the new model output. Default is `false`. |
```

| `max_prediction_tasks`  | Integer   | Optional  | The maximum number of concurrent model invocations that can run during document ingestion. Default is `10`. |
| `description` | String    | Optional  | A brief description of the processor. |
| `tag` | String    | Optional | An identifier tag for the processor. Useful for debugging to distinguish between processors of the same type. |

The `input_map` and `output_map` mappings support standard [JSON path](https://github.com/json-path/JsonPath) notation for specifying complex data structures. 
{: .note}

## Using the processor

Follow these steps to use the processor in a pipeline. You must provide a model ID when creating the processor. Before testing a pipeline or ingesting the documents using the processor, make sure that the model is successfully deployed. You can check the model state using the [Get Model API]({{site.url}}{{site.baseurl}}/ml-commons-plugin/api/model-apis/get-model/).

For local models, you must provide a `model_input` field that specifies the model input format. Add any input fields in `model_config` to `model_input`.

> For remote models, the `model_input` field is optional and its default value is `"{ \"parameters\": ${ml_inference.parameters} }`.

Reviewer 3: ```suggestion
For remote models, the `model_input` field is optional, and its default value is `"{ \"parameters\": ${ml_inference.parameters} }`.
```

| `model_id`                  | String    | Required                                                    | The ID of the ML model to be used by the processor.                                                                                                                                                                                                                                                                                                                                                         |
| `function_name`             | String    | Optional for remote models.<br/>Required for local models   | The function name of the ML model configured in the processor. The default value is `remote`                                                                                                                                                                                                                                                                                                                |

> | `input_map`                 | Array     | Optional                                                    | An array specifying how to map fields from the ingested document to the model's input. If no input mapping is specified, then all fields from the document are used as the model input by default. The `input_map` size indicates the number of prediction tasks. In each map, the key represents the model input field name, and the value specifies the document field used to construct the model input. |

Reviewer 2: @rbhavna Have you verified this is optional for both remote and local model?

Author: Sorry this is required for local models. Will update the documentation

```
{% include copy-curl.html %}

> For externally hosted models, in a Predict API request, all fields are usually nested inside the `parameters` object:

Reviewer 4: Usually, or will always be? From my understanding, it is always.

Reviewer 4: Since the ml inference processor makes this assumption within the mapping / conversion logic, I think we should make this a bit clearer.

Reviewer 4: cc @mingshl

Author: We are specifying here to the user that they need not give parameters path [here](https://github.com/opensearch-project/documentation-website/pull/7368/files#diff-b777c14a01b3fe5457f898d2f6a8d550f68c76ef49629d5fd436917c9c4080f8R131). Should we add more info?

```
{% include copy-curl.html %}

> For externally hosted models, in a Predict API request, all fields are usually nested inside the `parameters` object:

Reviewer 3: "For a Predict API request in an externally hosted model"? The back-to-back phrases are a little awkward.

> For externally hosted models, in a Predict API request, all fields are usually nested inside the `parameters` object:

Reviewer 5: ```suggestion
For a Predict API request to an externally hosted model, all fields are usually nested inside the `parameters` object:
```

```json
POST /_plugins/_ml/models/cleMb4kBJ1eYAeTMFFg4/_predict
{
  "parameters": {
    "input": [
      {
        ...
      }
    ]
  }
}
```

> When specifying the `input_map` for an externally hosted model, you can directly reference the `input` field  instead of providing its dot path `parameters.input`:

Reviewer 3: ```suggestion
When specifying the `input_map` for an externally hosted model, you can directly reference the `input` field instead of providing its dot path `parameters.input`:
```

{: .note}

### Example ingest processor configuration for local models

The following example creates an ingest pipeline for a `SENTENCE_TRANSFORMER` local model that requires an `text_docs` field, produces a `data` field for prediction in the model output path `$.inference_results.*.output.*.data`. The processor converts the `context` field into text embeddings and stores in the `context_embedding` field. The `input_map` and `output_map` mappings support standard JSON path notation for complex data structures.

For local models, the `model_input` field is mandatory and you should specify the inputs and their formats you want to provide the ML Model. If there are any inputs in `model_config` map to be used by model, they should also be added to `model_input`

For remote models, the `model_input` field is optional and its default value is `"{ \"parameters\": ${ml_inference.parameters} }`.
{: .note}

```json

PUT /_ingest/pipeline/ml_inference_pipeline_local
{
  "description": "ingest reviews and generate embedding",
  "processors": [
    {
      "ml_inference": {
        "function_name": "text_embedding",
        "full_response_path": true,

>         "model_id": "<your model id>",

Reviewer 2: Better to explain which model you are using , suggest use one of pretrained models

        "model_config": {
          "return_number": true
        },

>         "model_input": "{ \"text_docs\": ${input_map.text_docs}, \"return_number\": ${model_config.return_number} }",

Reviewer 2: Suggest also add `"target_response": ["sentence_embedding"]`

```
POST /_plugins/_ml/_predict/text_embedding/WWQI44MBbzI2oUKAvNUt
{
  "text_docs":[ "today is sunny", "hello world"],
  "return_number": true,
  "target_response": ["sentence_embedding"]
}
```

        "input_map": [
          {
            "text_docs": "book.*.chunk.text.*.context"
          }
        ],
        "output_map": [
          {

>             "book.*.chunk.text.*.context_embedding": "$.inference_results.*.output.*.dataAsMap.response"

Reviewer 2: Seems this is not for local model, which doesn't have dataAsMap

Once you have created an ingest pipeline, you need to create an index for ingestion and ingest documents into the index.
{: .note}

### Example: Local model

The following example configures an `ml_inference` processor with a local model.

**Step 1: Create a pipeline**

The following example creates an ingest pipeline for the `huggingface/sentence-transformers/all-distilroberta-v1` local model. The model is a sentence transformer [pretrained model]({{site.url}}{{site.baseurl}}/ml-commons-plugin/pretrained-models/#sentence-transformers) hosted in your OpenSearch cluster. 

> If you invoke the model using the Predict API, the request looks like this:

Reviewer 3: ```suggestion
If you invoke the model using the Predict API, then the request looks like this:
```

> If you invoke the model using the Predict API, the request looks like this:

Reviewer 3: "looks like this" => "appears as follows"?

```json
POST /_plugins/_ml/_predict/text_embedding/cleMb4kBJ1eYAeTMFFg4
{
  "text_docs":[ "today is sunny"],
  "return_number": true,
  "target_response": ["sentence_embedding"]
}
```

> Using this schema, you'll specify the `model_input` as follows:

Reviewer 3: ```suggestion
Using this schema, specify the `model_input` as follows:
```

```json
 "model_input": "{ \"text_docs\": ${input_map.text_docs}, \"return_number\": ${model_config.return_number}, \"target_response\": ${model_config.target_response} }"
```

> In the `input_map`, you'll map the document `book.*.chunk.text.*.context` field to the `text_docs` field expected by the model:

Reviewer 3: ```suggestion
In the `input_map`, map the `book.*.chunk.text.*.context` document field to the `text_docs` field expected by the model:
```

```json
"input_map": [
  {
    "text_docs": "book.*.chunk.text.*.context"
  }
]
```

> Because you're specifying the field to convert into embeddings as a JSON path, you need to set the `full_response_path` to `true` so that the full JSON document is parsed to obtain the input field:

Reviewer 3: ```suggestion
Because you specified the field to convert into embeddings as a JSON path, you need to set the `full_response_path` to `true` so that the full JSON document is parsed to obtain the input field:
```

```json
"full_response_path": true
```

> The documents you'll index will look as follows. The text in the `context` field will be used to generate embeddings:

Reviewer 3: ```suggestion
The documents you index will appear as follows. The text in the `context` field will be used to generate embeddings:
```

> An ingested document will look as follows. The text in the `context` field will be used to generate embeddings:

Author: `ingested` could be confusing because this is what user sends but finally ingested document is the output of the pipeline.

```json
"book": [
  {
    "chunk": {
      "text": [
        {
          "chapter": "first chapter",
          "context": "this is the first part"
        }
      ]
    }
  }
]
```

The Predict API request returns the following response:

```json
{
  "inference_results" : [
    {
      "output" : [
        {
          "name" : "sentence_embedding",
          "data_type" : "FLOAT32",
          "shape" : [
            768
          ],
          "data" : [
            0.25517133,
            -0.28009856,
            0.48519906,
            ...
          ]
        }
      ]
    }
  ]
}
```

> The model records embeddings in the `data` field at the output path `$.inference_results.*.output.*.data`. The `output_map` specifies to record the resulting embeddings in the `context_embedding` field of an ingested document at the same path as the `context` field:

Author: Can we simplify here as well? The output_map maps the model's embeddings in the path `$.inference_results.*.output.*.data` to the new ingest document field specified by the path `book.*.chunk.text.*.context_embedding`. Also it need not be in the same path as `context`. Depends on user's configuration of the processor

```json
"output_map": [
  {
    "book.*.chunk.text.*.context_embedding": "$.inference_results.*.output.*.data"
  }
]
```

To configure an `ml_inference` processor with a local model, specify the `function_name` explicitly. In this example, `function_name` is `text_embedding`. For information about valid `function_name` values, see [Configuration parameters](#configuration-parameters).

> The final configuration of the `ml_inference` processor with the local model in this example is as follows:

Reviewer 3: ```suggestion
In this example, the final configuration of the `ml_inference` processor with the local model is as follows:
```

        "ignore_failure": true
      }
    }
  ]
}
```
{% include copy-curl.html %}

**Step 2 (Optional): Test the pipeline**

To test the pipeline, run the following query:

```json
POST _ingest/pipeline/ml_inference_pipeline/_simulate
{
  "docs": [
    {
      "_index": "my_books",
      "_id": "1",
      "_source": {
        "book": [
          {
            "chunk": {
              "text": [
                {
                  "chapter": "first chapter",
                  "context": "this is the first part"
                },
                {
                  "chapter": "first chapter",
                  "context": "this is the second part"
                }
              ]
            }
          },
          {
            "chunk": {
              "text": [
                {
                  "chapter": "second chapter",
                  "context": "this is the third part"
                },
                {
                  "chapter": "second chapter",
                  "context": "this is the fourth part"
                }
              ]
            }
          }
        ]
      }
    }
  ]
}
```
{% include copy-curl.html %}

#### Response

> The response confirms that the processor has generated text embeddings in the `context_embedding` field. The document now contains both the `context` and `context_embedding` fields at the same path:

Reviewer 3: "in" or "on" the same path?

Once you have created an ingest pipeline, you need to create an index for ingestion and ingest documents into the index.
{: .note}

### Example: Local model

The following example configures an `ml_inference` processor with a local model.

**Step 1: Create a pipeline**

The following example creates an ingest pipeline for the `huggingface/sentence-transformers/all-distilroberta-v1` local model. The model is a sentence transformer [pretrained model]({{site.url}}{{site.baseurl}}/ml-commons-plugin/pretrained-models/#sentence-transformers) hosted in your OpenSearch cluster. 

If you invoke the model using the Predict API, the request looks like this:

```json
POST /_plugins/_ml/_predict/text_embedding/cleMb4kBJ1eYAeTMFFg4
{
  "text_docs":[ "today is sunny"],
  "return_number": true,
  "target_response": ["sentence_embedding"]
}
```

Using this schema, you'll specify the `model_input` as follows:

```json
 "model_input": "{ \"text_docs\": ${input_map.text_docs}, \"return_number\": ${model_config.return_number}, \"target_response\": ${model_config.target_response} }"
```

> The model converts the text in the `text_docs` field to embeddings so it expects an ingested document to contain a `text_docs` field. In this example, ingested documents do not contain a `text_docs` field. The `model_input` specifies to source the `text_docs` field from the `input_map`. The `input_map`, in turn, specifies that the input text resides at the `book.*.chunk.text.*.context` JSON path:

Author: Could you help simplify this part? We just say that we map `book.*.chunk.text.*.context` to model's input `text_docs` field. More details could be confusing sometime.
